libcaf_net/caf/net/datagram_transport.hpp

```diff
@@ -98,17 +98,16 @@ class datagram_transport : public datagram_transport_base<Factory> {
   }
 
   bool handle_write_event(endpoint_manager& manager) override {
-    CAF_LOG_TRACE(CAF_ARG(this->handle_.id)
+    CAF_LOG_TRACE(CAF_ARG2("handle", this->handle_.id)
                   << CAF_ARG2("queue-size", packet_queue_.size()));
-    // Try to write leftover data.
-    write_some();
-    // Get new data from parent.
-    for (auto msg = manager.next_message(); msg != nullptr;
-         msg = manager.next_message()) {
-      this->next_layer_.write_message(*this, std::move(msg));
-    }
-    // Write prepared data.
-    return write_some();
+    auto get_message = [&] {
+      if (auto msg = manager.next_message())
+        this->next_layer_.write_message(*this, std::move(msg));
+      return !packet_queue_.empty();
```

Comment:
This returns `false` after dequeueing the last element.

```c++
      if (auto msg = manager.next_message()) {
        this->next_layer_.write_message(*this, std::move(msg));
        return true;
      }
      return false;
```

Also, I find the lambda name misleading. This doesn't just *get* a message. It fetches the next message and writes it. `fetch_next_message`?

Reply:
> This returns false after dequeueing the last element.

Why though? Every message that is dequeued should eventually be moved into the packet_queue of the transport. Hence this will only return `false` if there is nothing left in the queue.
But I see that just returning true and false is more simple. :+1:

Reply:
Ah, the `packet_queue_` is where `write_message` is supposed to put the outgoing messages? But you can still have the case that (for whatever reason) the next layer drops a message. In this case, we still want to return true to try again on the next cycle.

Reply:
> Ah, the packet_queue_ is where write_message is supposed to put the outgoing messages?

Exactly. 
I have changed it locally already. Your proposed change is definitely better suited.